### PR TITLE
CheckBox renovation

### PR DESCRIPTION
**Pull request by ksercs**
- [x] add `readOnly` prop
- [x] fix `useInkRipple` prop
- [x] fix QUnit tests
- [x] move onValueChanged to preact wrapper
- [x] implement validation in preact wrapper
- [x] add jest tests
- [x] make jest test coverage 100%
- [x] check renovated Checkbox inside of non-renovated widgets

### CODE REVIEW

File: js/renovation/check-box.tsx [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+      onKeyDown={viewModel.onWidgetKeyDown}
+      rtlEnabled={viewModel.props.rtlEnabled}
+      tabIndex={viewModel.props.tabIndex}
+      visible={viewModel.props.visible}
+      width={viewModel.props.width}
+      {...viewModel.restAttributes} // eslint-disable-line react/jsx-props-no-spreading
+    >
+      {name
+        ? <input ref={viewModel.submitInputRef} type="hidden" value={`${viewModel.props.value}`} name={name} />
+        : <input ref={viewModel.submitInputRef} type="hidden" value={`${viewModel.props.value}`} />}
```
Comment (MaximKudriavtsev) [inline js/renovation/check-box.tsx]: Can we use the following here or something similar?

```jsx
<input ref={viewModel.submitInputRef} type="hidden" value={`${viewModel.props.value}`} {...name && { name}} />
```

File: js/renovation/check-box.tsx [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+  onInactive(event: Event): void {
+    const { useInkRipple } = this.props;
+
+    useInkRipple && this.inkRippleRef.hideWave({ element: this.widgetRef, event });
+  }
+
+  onWidgetClick(event: Event): void {
+    const { value, onValueChanged } = this.props;
+
+    this.props.value = !value;
```
Comment (MaximKudriavtsev) [inline js/renovation/check-box.tsx]: A strange string. According ideology we should not change props inside

Comment (ksercs) [reply js/renovation/check-box.tsx]: I agree, but how should we change the value in this case?

File: js/renovation/check-box.tsx (hunk 1 above)
Comment (ksercs) [reply js/renovation/check-box.tsx]: I think so. I'll try soon, thx

**Review by MikeVitik — COMMENTED** [review]
Maybe onWidgetKeyDown(options): void | ... {
with out 
return undefined;

**Comment by ksercs** [comment]
> Maybe onWidgetKeyDown(options): void | ... {
> with out
> return undefined;

I'll research it, thx. It's just copypaste from button :-)

File: js/renovation/check-box.tsx [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+
+    // eslint-disable-next-line  @typescript-eslint/no-non-null-assertion
+    onValueChanged!({
+      event,
+      element: this.submitInputRef.parentNode,
+      previousValue: value,
+      value: this.props.value,
```
Comment (LazyLahtak) [inline js/renovation/check-box.tsx]: We shouldn't use value of `TwoWay` property if we assigned it above. It doesn't change instantly.

Comment (ksercs) [reply js/renovation/check-box.tsx]: Excellent remark, thanks!

Comment (ksercs) [reply js/renovation/check-box.tsx]: fixed

File: js/renovation/check-box.tsx (hunk 1 above)
Comment (ksercs) [reply js/renovation/check-box.tsx]: fixed

**Comment by ksercs** [comment]
> > Maybe onWidgetKeyDown(options): void | ... {
> > with out
> > return undefined;
> 
> I'll research it, thx. It's just copypaste from button :-)

The reason of this strange code appearance is `consistent-return` eslint rule

File: js/renovation/check-box.tsx [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+  ComponentBindings,
+  JSXComponent,
+  Method,
+  OneWay,
+  TwoWay,
+  Ref,
+  Effect,
+} from 'devextreme-generator/component_declaration/common';
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+import { h } from 'preact';
```
Comment (roman-simionov) [inline js/renovation/check-box.tsx]: Don't forget to remove this import after a merge with 20_2
It is unnecessary after this commit
https://github.com/DevExpress/DevExtreme/commit/6a9f68a0ee92b6b62b81194e7198e0460132aa5c

Comment (ksercs) [reply js/renovation/check-box.tsx]: Thanks!

File: js/renovation/check-box.tsx [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+    useInkRipple && this.inkRippleRef.showWave({ element: this.iconRef, event, wave: 1 });
+  }
+
+  onInactive(event: Event): void {
+    const { useInkRipple } = this.props;
+
+    useInkRipple && this.inkRippleRef.hideWave({ element: this.iconRef, event, wave: 1 });
+  }
+
+  onFocusin(event: Event): void {
```
Comment (LazyLahtak) [inline js/renovation/check-box.tsx]: `onFocusin` -> `onFocusIn`

File: js/renovation/check-box.tsx [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+    const { useInkRipple, onFocusin } = this.props;
+
+    useInkRipple && this.inkRippleRef.showWave({ element: this.iconRef, event, wave: 0 });
+    onFocusin?.(event);
+  }
+
+  onFocusout(event: Event): void {
```
Comment (LazyLahtak) [inline js/renovation/check-box.tsx]: `onFocusOut` -> `onFocusOut`

File: js/renovation/widget.tsx [hunk 7]
```diff
@@ -157,56 +167,62 @@ export default class Widget extends JSXComponent(WidgetProps) {
...
+        (e) => { onFocusin?.(e); !e.isDefaultPrevented() && (this.focused = true); },
+        (e) => { onFocusout?.(e); !e.isDefaultPrevented() && (this.focused = false); },
         {
           isFocusable: focusable,
           namespace,
         });
-
-      return () => focus.off(this.widgetRef, { namespace });
+      // eslint-disable-next-line  @typescript-eslint/no-explicit-any
+      return (): any => focus.off(this.widgetRef, { namespace });
```
Comment (LazyLahtak) [inline js/renovation/widget.tsx]: `any` => `void`

File: js/renovation/widget.tsx [hunk 8]
```diff
@@ -217,58 +233,64 @@ export default class Widget extends JSXComponent(WidgetProps) {
         () => { !this.active && (this.hovered = true); },
         () => { this.hovered = false; },
         { selector, namespace });
-
-      return () => hover.off(this.widgetRef, { selector, namespace });
+      // eslint-disable-next-line  @typescript-eslint/no-explicit-any
+      return (): any => hover.off(this.widgetRef, { selector, namespace });
```
Comment (LazyLahtak) [inline js/renovation/widget.tsx]: `any` => `void`

File: js/renovation/widget.tsx [hunk 9]
```diff
@@ -217,58 +233,64 @@ export default class Widget extends JSXComponent(WidgetProps) {
...
+  keyboardEffect(): any {
     const { focusStateEnabled, onKeyDown } = this.props;
 
     if (focusStateEnabled || onKeyDown) {
+      // eslint-disable-next-line  @typescript-eslint/no-non-null-assertion
       const id = keyboard.on(this.widgetRef, this.widgetRef, (e) => onKeyDown!(e));
-
-      return () => keyboard.off(id);
+      // eslint-disable-next-line  @typescript-eslint/no-explicit-any
+      return (): any => keyboard.off(id);
```
Comment (LazyLahtak) [inline js/renovation/widget.tsx]: `any` => `void`

File: js/renovation/widget.tsx [hunk 10]
```diff
@@ -217,58 +233,64 @@ export default class Widget extends JSXComponent(WidgetProps) {
...
+  resizeEffect(): any {
     const namespace = `${this.props.name}VisibilityChange`;
     const { onDimensionChanged } = this.props;
 
     if (onDimensionChanged) {
       resize.on(this.widgetRef, onDimensionChanged, { namespace });
-
-      return () => resize.off(this.widgetRef, { namespace });
+      // eslint-disable-next-line  @typescript-eslint/no-explicit-any
+      return (): any => resize.off(this.widgetRef, { namespace });
```
Comment (LazyLahtak) [inline js/renovation/widget.tsx]: `any` => `void`

File: js/renovation/widget.tsx [hunk 11]
```diff
@@ -217,58 +233,64 @@ export default class Widget extends JSXComponent(WidgetProps) {
...
       visibility.on(this.widgetRef,
+        // eslint-disable-next-line  @typescript-eslint/no-non-null-assertion
         () => onVisibilityChange!(true),
+        // eslint-disable-next-line  @typescript-eslint/no-non-null-assertion
         () => onVisibilityChange!(false),
         { namespace });
-
-      return () => visibility.off(this.widgetRef, { namespace });
+      // eslint-disable-next-line  @typescript-eslint/no-explicit-any
+      return (): any => visibility.off(this.widgetRef, { namespace });
```
Comment (LazyLahtak) [inline js/renovation/widget.tsx]: `any` => `void`

File: js/renovation/widget.tsx [hunk 12]
```diff
@@ -157,56 +167,62 @@ export default class Widget extends JSXComponent(WidgetProps) {
...
+    } = this.props;
     const namespace = `${name}Focus`;
     const isFocusable = focusStateEnabled && !disabled;
 
     if (isFocusable) {
       focus.on(this.widgetRef,
-        (e) => { !e.isDefaultPrevented() && (this.focused = true); },
-        (e) => { !e.isDefaultPrevented() && (this.focused = false); },
+        (e) => { onFocusin?.(e); !e.isDefaultPrevented() && (this.focused = true); },
+        (e) => { onFocusout?.(e); !e.isDefaultPrevented() && (this.focused = false); },
```
Comment (LazyLahtak) [inline js/renovation/widget.tsx]: Should we call `onFocusIn` and `onFocusOut` if `e.isDefaultPrevented() === true`?

File: js/renovation/check-box.tsx (hunk 4 above)
Comment (ksercs) [reply js/renovation/check-box.tsx]: fixed

File: js/renovation/check-box.tsx (hunk 5 above)
Comment (ksercs) [reply js/renovation/check-box.tsx]: fixed

File: js/renovation/check-box.tsx (hunk 6 above)
Comment (ksercs) [reply js/renovation/check-box.tsx]: fixed

File: js/renovation/widget.tsx (hunk 7 above)
Comment (ksercs) [reply js/renovation/widget.tsx]: fixed

File: js/renovation/widget.tsx (hunk 8 above)
Comment (ksercs) [reply js/renovation/widget.tsx]: fixed

File: js/renovation/widget.tsx (hunk 9 above)
Comment (ksercs) [reply js/renovation/widget.tsx]: fixed

File: js/renovation/widget.tsx (hunk 10 above)
Comment (ksercs) [reply js/renovation/widget.tsx]: fixed

File: js/renovation/widget.tsx (hunk 11 above)
Comment (ksercs) [reply js/renovation/widget.tsx]: fixed

File: js/renovation/widget.tsx (hunk 12 above)
Comment (ksercs) [reply js/renovation/widget.tsx]: I think, we shouldn't. Fixed, thx

File: js/renovation/preact_wrapper/check_box.ts [hunk 13]
```diff
@@ -0,0 +1,251 @@
+/* eslint-disable no-underscore-dangle */
```
Comment (ksercs) [inline js/renovation/preact_wrapper/check_box.ts:1]: Most of the code is validation related and will be removed after [this card ](https://trello.com/c/bfMm4WY9)completion.

**Comment by ksercs** [comment]
@DokaRus, review the changes in `js/ui/list/ui.list.edit.decorator.selection.js` and `testing/tests/DevExpress.ui.widgets.editors/checkbox.tests.js`, please

File: js/renovation/ui/check_box.tsx [hunk 14]
```diff
@@ -0,0 +1,228 @@
...
+} from 'devextreme-generator/component_declaration/common';
+import { createDefaultOptionRules } from '../../core/options/utils';
+import devices from '../../core/devices';
+import { InkRipple } from './common/ink_ripple';
+import { Widget } from './common/widget';
+import * as themes from '../../ui/themes';
+import BaseComponent from '../preact_wrapper/check_box';
+import BaseWidgetProps from '../utils/base_props';
+
+const getCssClasses = (model: CheckBoxProps): string => {
```
Comment (MikeVitik) [inline js/renovation/ui/check_box.tsx:22]: Try to use combineClasses

File: js/renovation/ui/check_box.tsx [hunk 15]
```diff
@@ -0,0 +1,228 @@
...
+  contentReadyEffect(): void {
+    const { onContentReady } = this.props;
+    // eslint-disable-next-line  @typescript-eslint/no-non-null-assertion
+    onContentReady!({ element: this.widgetRef });
+  }
+
+  onActive(event: Event): void {
+    const { useInkRipple } = this.props;
+
+    useInkRipple && this.inkRippleRef.showWave({ element: this.iconRef, event, wave: 1 });
```
Comment (MikeVitik) [inline js/renovation/ui/check_box.tsx]: const { useInkRipple } = this.props;

    useInkRipple && this.inkRippleRef.showWave({ element: this.iconRef, event, wave: 1 });

May be extract to function wave(type: 'showWave' | 'hideWave', waveId: number)

File: js/renovation/ui/__tests__/check_box.test.tsx [hunk 16]
```diff
@@ -0,0 +1,478 @@
...
+
+  describe('Logic', () => {
+    describe('Getters', () => {
+      describe('aria', () => {
+        it('should have role = "checkbox"', () => {
+          expect(new CheckBox({}).aria).toMatchObject({ role: 'checkbox' });
+        });
+
+        // TODO: replace null with undefined after generator bug fix
+        [true, false, null].forEach((value) => {
```
Comment (MikeVitik) [inline js/renovation/ui/__tests__/check_box.test.tsx]: https://github.com/facebook/jest/tree/master/packages/jest-each

File: js/renovation/ui/check_box.tsx (hunk 14 above)
Comment (ksercs) [reply js/renovation/ui/check_box.tsx:22]: fixed

File: js/renovation/ui/check_box.tsx (hunk 15 above)
Comment (ksercs) [reply js/renovation/ui/check_box.tsx]: fixed

File: js/renovation/ui/__tests__/check_box.test.tsx (hunk 16 above)
Comment (ksercs) [reply js/renovation/ui/__tests__/check_box.test.tsx]: fixed

File: testing/tests/DevExpress.ui.widgets.editors/checkbox.tests.js [hunk 17]
```diff
@@ -2,6 +2,11 @@ import $ from 'jquery';
 import devices from 'core/devices';
 import keyboardMock from '../../helpers/keyboardMock.js';
 import { validateGroup } from 'ui/validation_engine';
+import registerComponent from 'core/component_registrator.js';
+import dxrCheckBox from 'renovation/ui/check_box.j.js';
+import dxCheckBox from 'ui/check_box';
```
Comment (DokaRus) [inline testing/tests/DevExpress.ui.widgets.editors/checkbox.tests.js:7]: Remove [R12](https://github.com/DevExpress/DevExtreme/pull/13644/files#diff-3e56680bc5b493b77956ddd3ca619633R12)
Double `import 'ui/check_box'`

File: js/ui/list/ui.list.edit.decorator.selection.js [hunk 18]
```diff
@@ -49,19 +49,20 @@ registerDecorator(
...
                 value: this._isSelected($itemElement),
                 focusStateEnabled: false,
                 hoverStateEnabled: false,
                 onValueChanged: (function(e) {
                     this._processCheckedState($itemElement, e.value);
                     e.event && e.event.stopPropagation();
-                }).bind(this)
+                }).bind(this),
+                elementAttr: {
+                    class: this._controlClass
```
Comment (DokaRus) [inline js/ui/list/ui.list.edit.decorator.selection.js]: What's it done for?

File: testing/tests/DevExpress.ui.widgets.editors/checkbox.tests.js (hunk 17 above)
Comment (ksercs) [reply testing/tests/DevExpress.ui.widgets.editors/checkbox.tests.js:7]: fixed

File: js/ui/list/ui.list.edit.decorator.selection.js (hunk 18 above)
Comment (ksercs) [reply js/ui/list/ui.list.edit.decorator.selection.js]: It's generator bug WA. I don't wanna remove it on the off-chance, to make it easier to find it in the future. Can it break smthing, how do u think?

File: js/renovation/ui/common/widget.tsx [hunk 19]
```diff
@@ -112,7 +119,8 @@ export class Widget extends JSXComponent(WidgetProps) {
   widgetRef!: HTMLDivElement;
 
   @Effect()
-  accessKeyEffect() {
+  // eslint-disable-next-line  @typescript-eslint/no-explicit-any
+  accessKeyEffect(): any {
```
Comment (LazyLahtak) [inline js/renovation/ui/common/widget.tsx]: Maybe we can use `(() => void) | undefined`?

Also, we can try replace `return undefined` by `return noop` and only `() => void` will remain.

File: js/renovation/ui/common/widget.tsx [hunk 20]
```diff
@@ -126,14 +134,16 @@ export class Widget extends JSXComponent(WidgetProps) {
         }
       }, { namespace });
 
-      return () => dxClick.off(this.widgetRef, { namespace });
+      // eslint-disable-next-line  @typescript-eslint/no-explicit-any
+      return (): any => dxClick.off(this.widgetRef, { namespace });
```
Comment (LazyLahtak) [inline js/renovation/ui/common/widget.tsx]: `any` => `void`

File: js/ui/list/ui.list.edit.decorator.selection.js (hunk 18 above)
Comment (DokaRus) [reply js/ui/list/ui.list.edit.decorator.selection.js]: I don't think so. Just looks weird :)
Could you add a comment note to remember to remove this WA in the future?

Comment (ksercs) [reply js/ui/list/ui.list.edit.decorator.selection.js]: I decided to backup this WA locally and not push to the repo :-) It's will be safer

File: js/renovation/ui/common/widget.tsx (hunk 19 above)
Comment (ksercs) [reply js/renovation/ui/common/widget.tsx]: fixed

File: js/renovation/ui/common/widget.tsx (hunk 20 above)
Comment (ksercs) [reply js/renovation/ui/common/widget.tsx]: fixed

File: js/renovation/ui/common/widget.tsx [hunk 21]
```diff
@@ -157,56 +168,72 @@ export class Widget extends JSXComponent(WidgetProps) {
...
     }
 
-    return undefined;
+    return noop;
   }
 
   @Effect()
-  clickEffect() {
-    const { name, onClick } = this.props;
+  // eslint-disable-next-line  @typescript-eslint/no-explicit-any
```
Comment (LazyLahtak) [inline js/renovation/ui/common/widget.tsx]: I think we don't need this anymore.

Comment (ksercs) [reply js/renovation/ui/common/widget.tsx]: fixed

**Review by MikeVitik — DISMISSED** [review]

**Comment by San4es** [comment]
https://devextreme-ci.devexpress.com/DevExpress/DevExtreme/34869/131
https://github.com/DevExpress/DevExtreme/pull/13644/checks?check_run_id=899242653

File: js/renovation/ui/button.tsx [hunk 22]
```diff
@@ -106,18 +107,21 @@ export class ButtonProps extends BaseWidgetProps {
   @Event({
     actionConfig: { excludeValidators: ['readOnly'] },
   })
-  onClick?: (e: any) => any = noop;
 
-  @Event() onSubmit?: (e: any) => any = noop;
+  // eslint-disable-next-line  @typescript-eslint/no-explicit-any
+  onClick?: (e: any) => void = noop;
```
Comment (LazyLahtak) [inline js/renovation/ui/button.tsx]: If we want to get rid of `any`, I think we can use somethink like:
```tsx
onClick?: (e: {
    event: Event;
    validationGroup?: string;
  }) => any = noop;
```

File: js/renovation/ui/button.tsx [hunk 23]
```diff
@@ -106,18 +107,21 @@ export class ButtonProps extends BaseWidgetProps {
...
+
+  // eslint-disable-next-line  @typescript-eslint/no-explicit-any
+  @Event() onSubmit?: (e: any) => void = noop;
```
Comment (LazyLahtak) [inline js/renovation/ui/button.tsx]: Same as above

**Review by DokaRus — APPROVED** [review]

**Review by churkin — APPROVED** [review]

**Review by LazyLahtak — APPROVED** [review]

**Review by roman-simionov — APPROVED** [review]

**Pull request merged**